File: biodesign-library/java/BioDesignMetadataLibrary.java
import com.genbio.genBio.BioDesignMetadata.diff_match_patch;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;

import javax.crypto.Cipher;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.lang.reflect.Type;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.security.MessageDigest;
import java.security.SecureRandom;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.*;

public class BioDesignMetadataLibrary {
    private static final DateTimeFormatter TS_FORMAT = DateTimeFormatter.ofPattern("MM/dd/yyyy, HH:mm:ss");
    private static final Gson GSON = new GsonBuilder().setPrettyPrinting().create();

    public BioDesignMetadata createMetadata(
            String parentMetadataId,
            String designName,
            String author,
            String description,
            String design
    ) throws Exception {
        BioDesignMetadata metadata = new BioDesignMetadata();
        metadata.setId(UUID.randomUUID().toString());
        metadata.setParentMetadataId(parentMetadataId);
        metadata.setDesignName(designName);
        metadata.setDesignChecksum(calculateChecksum(design));
        metadata.setAuthor(author);
        metadata.setDescription(description);
        metadata.setLastUpdated(nowTimestamp());
        metadata.setChangelog(new ArrayList<>());

        Files.writeString(
                Path.of("library", "metadata_" + designName + ".json"),
                GSON.toJson(metadata),
                StandardCharsets.UTF_8
        );
        return metadata;
    }

    public BioDesignMetadata updateMetadataWithOperation(
            String metadataPath,
            String design,
            String operationCode,
            Map<String, Object> operationDetails,
            String change,
            List<Map<String, String>> comments,
            String status
    ) throws Exception {
        BioDesignMetadata metadata = GSON.fromJson(Files.readString(Path.of(metadataPath), StandardCharsets.UTF_8), BioDesignMetadata.class);
        if (metadata.getChangelog() != null) {
            for (BioDesignOperation operation : metadata.getChangelog()) {
                if (operation.getComments() == null) {
                    operation.setComments(new ArrayList<>());
                }
                if (operation.getStatus() == null) {
                    operation.setStatus("");
                }
            }
        }

        metadata.setLastUpdated(nowTimestamp());
        metadata.setDesignChecksum(calculateChecksum(design));

        BioDesignOperation operation = new BioDesignOperation();
        operation.setOperationCode(operationCode);
        operation.setOperationDetails(operationDetails);
        operation.setChange(change);
        operation.setTimestamp(metadata.getLastUpdated());
        operation.setTool("BioDesign tool");
        operation.setComments(comments != null ? comments : new ArrayList<>());
        operation.setStatus(status != null ? status : "");
        metadata.getChangelog().add(operation);

        Files.writeString(Path.of(metadataPath), GSON.toJson(metadata), StandardCharsets.UTF_8);
        return metadata;
    }

    public BioDesignMetadata updateMetadataWithOperation(
            String metadataPath,
            String design,
            String operationCode,
            Map<String, Object> operationDetails,
            String change
    ) throws Exception {
        return updateMetadataWithOperation(metadataPath, design, operationCode, operationDetails, change, new ArrayList<>(), "");
    }

    public static String calculateChecksum(String inputStr) throws Exception {
        MessageDigest digest = MessageDigest.getInstance("SHA-256");
        byte[] hashBytes = digest.digest((inputStr == null ? "" : inputStr.toLowerCase()).getBytes(StandardCharsets.UTF_8));
        StringBuilder hexString = new StringBuilder();
        for (byte b : hashBytes) {
            hexString.append(String.format("%02x", b));
        }
        return hexString.toString();
    }

    public static String computeDifference(String originalString, String newString) {
        diff_match_patch dmp = new diff_match_patch();
        LinkedList<diff_match_patch.Patch> patches = dmp.patch_make(newString, originalString);
        return dmp.patch_toText(patches);
    }

    public static List<Map<String, Object>> computeRevisions(String lastDesign, List<Map<String, Object>> changelog) {
        diff_match_patch dmp = new diff_match_patch();
        int currentRevision = changelog.size();
        List<Map<String, Object>> revisions = new ArrayList<>();
        List<Map<String, Object>> reversed = new ArrayList<>(changelog);
        Collections.reverse(reversed);
        String currentDesign = lastDesign;

        for (Map<String, Object> operation : reversed) {
            if (!operation.containsKey("comments")) {
                operation.put("comments", new ArrayList<>());
            }
            if (!operation.containsKey("status")) {
                operation.put("status", "");
            }

            Map<String, Object> revision = new LinkedHashMap<>();
            revision.put("revision", currentRevision);
            revision.put("design", currentDesign);
            revision.putAll(operation);
            revisions.add(revision);

            Object changeObj = operation.get("change");
            String change = changeObj == null ? "" : changeObj.toString();
            if (!change.isEmpty()) {
                List<diff_match_patch.Patch> patches = dmp.patch_fromText(change);
                Object[] result = dmp.patch_apply(new LinkedList<>(patches), currentDesign);
                currentDesign = (String) result[0];
            }
            currentRevision -= 1;
        }
        return revisions;
    }

    public static String decryptString(String encryptedBase64String) {
        String encryptionKey = System.getenv("BMDE_ENCRYPTION_KEY");
        if (encryptionKey == null || encryptionKey.isEmpty()) {
            throw new RuntimeException(
                    "BMDE_ENCRYPTION_KEY environment variable is not set. Please set it to decrypt metadata files."
            );
        }
        try {
            byte[] key = encryptionKey.getBytes(StandardCharsets.UTF_8);
            byte[] encryptedDataWithIv = Base64.getDecoder().decode(encryptedBase64String);
            byte[] iv = Arrays.copyOfRange(encryptedDataWithIv, 0, 16);
            byte[] encryptedData = Arrays.copyOfRange(encryptedDataWithIv, 16, encryptedDataWithIv.length);

            Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
            cipher.init(Cipher.DECRYPT_MODE, new SecretKeySpec(key, "AES"), new IvParameterSpec(iv));
            byte[] originalData = cipher.doFinal(encryptedData);
            return new String(originalData, StandardCharsets.UTF_8);
        } catch (Exception e) {
            throw new RuntimeException(
                    "Failed to decrypt metadata. The encryption key may be incorrect or the data may be corrupted. " + e,
                    e
            );
        }
    }

    public static String encryptString(String stringToEncrypt) {
        String encryptionKey = System.getenv("BMDE_ENCRYPTION_KEY");
        if (encryptionKey == null || encryptionKey.isEmpty()) {
            throw new RuntimeException("BMDE_ENCRYPTION_KEY environment variable is not set.");
        }
        try {
            byte[] key = encryptionKey.getBytes(StandardCharsets.UTF_8);
            byte[] iv = new byte[16];
            new SecureRandom().nextBytes(iv);

            Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
            cipher.init(Cipher.ENCRYPT_MODE, new SecretKeySpec(key, "AES"), new IvParameterSpec(iv));
            byte[] encryptedData = cipher.doFinal(stringToEncrypt.getBytes(StandardCharsets.UTF_8));
            byte[] encryptedDataWithIv = new byte[iv.length + encryptedData.length];
            System.arraycopy(iv, 0, encryptedDataWithIv, 0, iv.length);
            System.arraycopy(encryptedData, 0, encryptedDataWithIv, iv.length, encryptedData.length);
            return Base64.getEncoder().encodeToString(encryptedDataWithIv);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    @SuppressWarnings("unchecked")
    public static List<Map<String, Object>> parseChangelogFromMetadataJson(String metadataJson) {
        Type mapType = new TypeToken<Map<String, Object>>() {}.getType();
        Map<String, Object> parsed = GSON.fromJson(metadataJson, mapType);
        Object changelogObj = parsed.get("changelog");
        if (!(changelogObj instanceof List)) return new ArrayList<>();
        return (List<Map<String, Object>>) changelogObj;
    }

    private static String nowTimestamp() {
        return LocalDateTime.now().format(TS_FORMAT);
    }
}

File: biodesign-library/typescript/BioDesignMetadataLibrary.ts
/**
 * Core BMDE library - metadata creation, encryption, checksum, and revision computation.
 */

import * as crypto from 'crypto';
import { createCipheriv, createDecipheriv, randomBytes } from 'crypto';
import { diff_match_patch } from 'diff-match-patch';
import * as dotenv from 'dotenv';
import * as fs from 'fs';

dotenv.config();

export interface BioDesignOperation {
  operationCode: string;
  operationDetails: Record<string, any>;
  change: string;
  timestamp: string;
  tool: string;
  comments: Array<Record<string, string>>;
  status: string;
}

export interface BioDesignMetadata {
  id: string;
  parentMetadataId: string | null;
  designName: string | null;
  designChecksum: string;
  author: string;
  description: string;
  lastUpdated: string;
  changelog: BioDesignOperation[];
}

function nowTimestamp(): string {
  const d = new Date();
  const m = String(d.getMonth() + 1).padStart(2, '0');
  const day = String(d.getDate()).padStart(2, '0');
  const y = d.getFullYear();
  const h = String(d.getHours()).padStart(2, '0');
  const min = String(d.getMinutes()).padStart(2, '0');
  const s = String(d.getSeconds()).padStart(2, '0');
  return `${m}/${day}/${y}, ${h}:${min}:${s}`;
}

export class BioDesignMetadataLibrary {
  createMetadata(
    parentMetadataId: string | null,
    designName: string,
    author: string,
    description: string,
    design: string
  ): BioDesignMetadata {
    const metadata: BioDesignMetadata = {
      id: crypto.randomUUID(),
      parentMetadataId: parentMetadataId ?? '',
      designName,
      designChecksum: this.calculateChecksum(design),
      author,
      description,
      lastUpdated: nowTimestamp(),
      changelog: [],
    };
    fs.writeFileSync(
      `library/metadata_${designName}.json`,
      JSON.stringify(metadata, null, 4),
      'utf-8'
    );
    return metadata;
  }

  updateMetadataWithOperation(
    metadataPath: string,
    design: string,
    operationCode: string,
    operationDetails: Record<string, any>,
    change: string,
    comments: Array<Record<string, string>> = [],
    status = ''
  ): BioDesignMetadata {
    const metadataJson = JSON.parse(fs.readFileSync(metadataPath, 'utf-8')) as BioDesignMetadata;
    if (metadataJson.changelog) {
      metadataJson.changelog.forEach((op: Partial<BioDesignOperation>) => {
        if (!op.comments) op.comments = [];
        if (typeof op.status !== 'string') op.status = '';
      });
    }
    const metadata: BioDesignMetadata = metadataJson;
    metadata.lastUpdated = nowTimestamp();
    metadata.designChecksum = this.calculateChecksum(design);
    metadata.changelog.push({
      operationCode,
      operationDetails,
      change,
      timestamp: metadata.lastUpdated,
      tool: 'BioDesign tool',
      comments,
      status: status || '',
    });
    fs.writeFileSync(metadataPath, JSON.stringify(metadata, null, 4), 'utf-8');
    return metadata;
  }

  calculateChecksum(inputStr: string): string {
    return calculateChecksum(inputStr);
  }

  computeDifference(originalString: string, newString: string): string {
    return computeDifference(originalString, newString);
  }

  computeRevisions(lastDesign: string, changelog: Array<Record<string, any>>): Array<Record<string, any>> {
    return computeRevisions(lastDesign, changelog);
  }

  decryptString(encryptedBase64String: string): string {
    return decryptString(encryptedBase64String);
  }

  encryptString(stringToEncrypt: string): string {
    return encryptString(stringToEncrypt);
  }
}

export function calculateChecksum(inputStr: string): string {
  const hash = crypto.createHash('sha256');
  hash.update((inputStr || '').toLowerCase(), 'utf-8');
  return hash.digest('hex');
}

export function computeDifference(originalString: string, newString: string): string {
  const dmp = new diff_match_patch();
  const patches = dmp.patch_make(newString, originalString);
  return dmp.patch_toText(patches);
}

export function computeRevisions(lastDesign: string, changelog: Array<Record<string, any>>): Array<Record<string, any>> {
  const dmp = new diff_match_patch();
  let currentRevision = changelog.length;
  const revisions: Array<Record<string, any>> = [];
  const reversed = [...changelog].reverse();
  let currentDesign = lastDesign;

  for (const operation of reversed) {
    if (!('comments' in operation)) operation.comments = [];
    if (!('status' in operation)) operation.status = '';

    revisions.push({
      revision: currentRevision,
      design: currentDesign,
      operationCode: operation.operationCode,
      operationDetails: operation.operationDetails,
      change: operation.change,
      timestamp: operation.timestamp,
      tool: operation.tool,
      comments: operation.comments,
      status: operation.status,
    });

    if (operation.change) {
      const patches = dmp.patch_fromText(operation.change);
      const [patched] = dmp.patch_apply(patches, currentDesign);
      currentDesign = patched;
    }
    currentRevision -= 1;
  }
  return revisions;
}

export function decryptString(encryptedBase64String: string): string {
  const encryptionKey = process.env.BMDE_ENCRYPTION_KEY;
  if (!encryptionKey) {
    throw new Error(
      'BMDE_ENCRYPTION_KEY environment variable is not set. Please set it to decrypt metadata files.'
    );
  }
  try {
    const key = Buffer.from(encryptionKey, 'utf-8');
    const encryptedDataWithIv = Buffer.from(encryptedBase64String, 'base64');
    const iv = encryptedDataWithIv.subarray(0, 16);
    const encryptedData = encryptedDataWithIv.subarray(16);
    const decipher = createDecipheriv('aes-256-cbc', key, iv);
    const decrypted = Buffer.concat([decipher.update(encryptedData), decipher.final()]);
    return decrypted.toString('utf-8');
  } catch (e: any) {
    throw new Error(
      `Failed to decrypt metadata. The encryption key may be incorrect or the data may be corrupted. ${e?.message || e}`
    );
  }
}

export function encryptString(stringToEncrypt: string): string {
  const encryptionKey = process.env.BMDE_ENCRYPTION_KEY;
  if (!encryptionKey) {
    throw new Error('BMDE_ENCRYPTION_KEY environment variable is not set.');
  }
  const key = Buffer.from(encryptionKey, 'utf-8');
  const iv = randomBytes(16);
  const cipher = createCipheriv('aes-256-cbc', key, iv);
  const encrypted = Buffer.concat([cipher.update(stringToEncrypt, 'utf-8'), cipher.final()]);
  return Buffer.concat([iv, encrypted]).toString('base64');
}
